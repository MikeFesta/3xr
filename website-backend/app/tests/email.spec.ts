// SPDX-License-Identifier: Apache-2.0
import Email from '../email/email';
import { TemplateRequest } from '@types';
import { EmailType } from '@enums';

describe('Email service:', () => {
  describe('* template generation', () => {
    let user: any;
    let product: any;
    let project: any;
    beforeEach(() => {
      user = { email: 'user@user.com', firstName: 'First', lastName: 'Last', emailNotifications: true };
      product = { name: 'productName', href: 'http://www.product.link' };
      project = { name: 'projectName', href: 'http://www.project.link' };
    });
    it('should correctly generate CLIENT_QA email', async () => {
      const templateRequest = await Email.generateTemplate({
        user,
        product,
        type: EmailType.CLIENT_QA,
      });
      expect(templateRequest?.email).toEqual('user@user.com');
      expect(templateRequest?.html).toMatchSnapshot();
    });
    it('should correctly generate JOB_IN_REVISION email', async () => {
      const templateRequest = await Email.generateTemplate({
        user,
        product,
        type: EmailType.JOB_IN_REVISION,
      });
      expect(templateRequest?.email).toEqual('user@user.com');
      expect(templateRequest?.html).toMatchSnapshot();
    });
    it('should correctly generate COMMENT email', async () => {
      const templateRequest = await Email.generateTemplate({
        user,
        product,
        type: EmailType.COMMENT,
        commentatorName: 'Commentator User',
      });
      expect(templateRequest?.email).toEqual('user@user.com');
      expect(templateRequest?.html).toMatchSnapshot();
    });
    it('should correctly generate PO_IN_PROGRESS email', async () => {
      const templateRequest = await Email.generateTemplate({
        user,
        project,
        type: EmailType.PO_IN_PROGRESS,
      });
      expect(templateRequest?.email).toEqual('user@user.com');
      expect(templateRequest?.html).toMatchSnapshot();
    });
    it('should correctly assign bcc', async () => {
      const templateRequest = await Email.generateTemplate({
        user,
        project,
        type: EmailType.PO_IN_PROGRESS,
        bcc: ['test@email.com', 'test2@email.com'],
      });
      expect(templateRequest?.bcc).toEqual(['test@email.com', 'test2@email.com']);
    });
    it('should correctly generate multiple emails', async () => {
      const requests: TemplateRequest[] = [
        {
          type: EmailType.COMMENT,
          user: { email: 'first@email.com', firstName: 'first', lastName: 'last', emailNotifications: true },
          product,
          commentatorName: 'Commentator User',
        },
        {
          type: EmailType.JOB_IN_REVISION,
          user: { email: 'second@email.com', firstName: 'first', lastName: 'last', emailNotifications: true },
          product,
        },
      ];
      const templateRequests = await Email.generateTemplates(requests);
      if (templateRequests) {
        templateRequests.forEach((templateRequest, idx) => {
          expect(templateRequest?.email).toEqual(requests[idx].user.email);
          expect(templateRequest?.html).toMatchSnapshot();
        });
      }
    });
    it('should not generate emails for users with disabled emailNotifications', async () => {
      const requests: TemplateRequest[] = [
        {
          type: EmailType.COMMENT,
          user: { email: 'first@email.com', firstName: 'first', lastName: 'last', emailNotifications: false },
          product,
          commentatorName: 'Commentator User',
        },
        {
          type: EmailType.JOB_IN_REVISION,
          user: { email: 'second@email.com', firstName: 'first', lastName: 'last', emailNotifications: true },
          product,
        },
      ];

      const templateRequests = await Email.generateTemplates(requests);

      if (templateRequests) {
        expect(templateRequests?.length).toEqual(1);
        expect(templateRequests[0].email).toBe('second@email.com');
      } else {
        console.log(templateRequests);
        throw new Error('template requests were not generated');
      }
    });
  });
});
