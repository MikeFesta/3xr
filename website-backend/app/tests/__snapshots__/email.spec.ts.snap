// SPDX-License-Identifier: Apache-2.0
// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email service: * template generation should correctly generate CLIENT_QA email 1`] = `
<!DOCTYPE html>
<html lang="en">

  <head>
    <title>3XR</title>
  </head>

  <body style="font-family: Poppins, Verdana, Geneva, Tahoma, sans-serif;">
    <div id="main" style="width:100%;">
      <div data-server-rendered="true">
        <div style="background-color: #0b82c5; text-align:left;  padding:2%">
          <img width="120px" src="https://cdn.3xr.com/images/3xr-email.png" style="float: right; padding-right:20px;">
          <h1 style="color:#FFF;padding-left:8%; padding-top:20px; padding-bottom: 20px;">Asset Ready For Review</h1>
        </div>
        <div style=" padding-right:10%; padding-left:10%; padding-top:20px;">
          <p>productName is ready to review - Log in here to QA your model</p>
          <p style="padding-top:15px; padding-bottom:15px;">
            <a href="http://www.product.link" style="color:#FFF; cursor:pointer;">
              <button style="cursor:pointer;background-color:#dda349; color=:#FFF; border:none; border-radius:5px;  padding-left:15px;  padding-right:15px; padding-top:5px; padding-bottom:5px; font-size:20px">Review asset</button>
            </a>
          </p>
          <div style="border-bottom:lightgray solid 2px"></div>
          <p>Have questions? Email
            <a href="mailto:sales@3xr.com" style="color: #0b82c5;">sales@3xr.com</a>
          </p>
        </div>
      </div>
    </div>
  </body>

</html>
`;

exports[`Email service: * template generation should correctly generate COMMENT email 1`] = `
<!DOCTYPE html>
<html lang="en">

  <head>
    <title>3XR</title>
  </head>

  <body style="font-family: Poppins, Verdana, Geneva, Tahoma, sans-serif;">
    <div id="main" style="width:100%;">
      <div data-server-rendered="true">
        <div style="background-color: #0b82c5; text-align:left;  padding:2%">
          <img width="120px" src="https://cdn.3xr.com/images/3xr-email.png" style="float: right; padding-right:20px;">
          <h1 style="color:#FFF;padding-left:8%; padding-top:20px; padding-bottom: 20px;">A Comment Has Been Made On Your Product</h1>
        </div>
        <div style=" padding-right:10%; padding-left:10%; padding-top:20px;">
          <p>Commentator User left a comment on productName, please review at your
            convenience.</p>
          <p style="padding-top:15px; padding-bottom:15px;">
            <a href="http://www.product.link" style="color:#FFF; cursor:pointer;">
              <button style="cursor:pointer;background-color:#dda349; color=:#FFF; border:none; border-radius:5px;  padding-left:15px;  padding-right:15px; padding-top:5px; padding-bottom:5px; font-size:20px">View comment</button>
            </a>
          </p>
          <div style="border-bottom:lightgray solid 2px"></div>
          <p>Have questions? Email
            <a href="mailto:sales@3xr.com" style="color: #0b82c5;">sales@3xr.com</a>
          </p>
        </div>
      </div>
    </div>
  </body>

</html>
`;

exports[`Email service: * template generation should correctly generate JOB_IN_REVISION email 1`] = `
<!DOCTYPE html>
<html lang="en">

  <head>
    <title>3XR</title>
  </head>

  <body style="font-family: Poppins, Verdana, Geneva, Tahoma, sans-serif;">
    <div id="main" style="width:100%;">
      <div data-server-rendered="true">
        <div style="background-color: #0b82c5; text-align:left;  padding:2%">
          <img width="120px" src="https://cdn.3xr.com/images/3xr-email.png" style="float: right; padding-right:20px;">
          <h1 style="color:#FFF;padding-left:8%; padding-top:20px; padding-bottom: 20px;">Revision Requested</h1>
        </div>
        <div style=" padding-right:10%; padding-left:10%; padding-top:20px;">
          <p>productName has been sent back to our artists with your feedback, we will
            notify you again when your requested revisions have been made. Please expect
            a 48 hour turnaround.</p>
          <p style="padding-top:15px; padding-bottom:15px;">
            <a href="http://www.product.link" style="color:#FFF; cursor:pointer;">
              <button style="cursor:pointer;background-color:#dda349; color=:#FFF; border:none; border-radius:5px;  padding-left:15px;  padding-right:15px; padding-top:5px; padding-bottom:5px; font-size:20px">Review asset</button>
            </a>
          </p>
          <div style="border-bottom:lightgray solid 2px"></div>
          <p>Have questions? Email
            <a href="mailto:sales@3xr.com" style="color: #0b82c5;">sales@3xr.com</a>
          </p>
        </div>
      </div>
    </div>
  </body>

</html>
`;

exports[`Email service: * template generation should correctly generate PO_IN_PROGRESS email 1`] = `
<!DOCTYPE html>
<html lang="en">

  <head>
    <title>3XR</title>
  </head>

  <body style="font-family: Poppins, Verdana, Geneva, Tahoma, sans-serif;">
    <div id="main" style="width:100%;">
      <div data-server-rendered="true">
        <div style="background-color: #0b82c5; text-align:left;  padding:2%">
          <img width="120px" src="https://cdn.3xr.com/images/3xr-email.png" style="float: right; padding-right:20px;">
          <h1 style="color:#FFF;padding-left:8%; padding-top:20px; padding-bottom: 20px;">Purchase Order Is In Progress</h1>
        </div>
        <div style=" padding-right:10%; padding-left:10%; padding-top:20px;">
          <p>projectName has begun production with 3XR artists. From here you will
            receive updates as your models become ready for your review.</p>
          <p style="padding-top:15px; padding-bottom:15px;">
            <a href="http://www.project.link" style="color:#FFF; cursor:pointer;">
              <button style="cursor:pointer;background-color:#dda349; color=:#FFF; border:none; border-radius:5px;  padding-left:15px;  padding-right:15px; padding-top:5px; padding-bottom:5px; font-size:20px">Review project</button>
            </a>
          </p>
          <div style="border-bottom:lightgray solid 2px"></div>
          <p>Have questions? Email
            <a href="mailto:sales@3xr.com" style="color: #0b82c5;">sales@3xr.com</a>
          </p>
        </div>
      </div>
    </div>
  </body>

</html>
`;

exports[`Email service: * template generation should correctly generate multiple emails 1`] = `
<!DOCTYPE html>
<html lang="en">

  <head>
    <title>3XR</title>
  </head>

  <body style="font-family: Poppins, Verdana, Geneva, Tahoma, sans-serif;">
    <div id="main" style="width:100%;">
      <div data-server-rendered="true">
        <div style="background-color: #0b82c5; text-align:left;  padding:2%">
          <img width="120px" src="https://cdn.3xr.com/images/3xr-email.png" style="float: right; padding-right:20px;">
          <h1 style="color:#FFF;padding-left:8%; padding-top:20px; padding-bottom: 20px;">A Comment Has Been Made On Your Product</h1>
        </div>
        <div style=" padding-right:10%; padding-left:10%; padding-top:20px;">
          <p>Commentator User left a comment on productName, please review at your
            convenience.</p>
          <p style="padding-top:15px; padding-bottom:15px;">
            <a href="http://www.product.link" style="color:#FFF; cursor:pointer;">
              <button style="cursor:pointer;background-color:#dda349; color=:#FFF; border:none; border-radius:5px;  padding-left:15px;  padding-right:15px; padding-top:5px; padding-bottom:5px; font-size:20px">View comment</button>
            </a>
          </p>
          <div style="border-bottom:lightgray solid 2px"></div>
          <p>Have questions? Email
            <a href="mailto:sales@3xr.com" style="color: #0b82c5;">sales@3xr.com</a>
          </p>
        </div>
      </div>
    </div>
  </body>

</html>
`;

exports[`Email service: * template generation should correctly generate multiple emails 2`] = `
<!DOCTYPE html>
<html lang="en">

  <head>
    <title>3XR</title>
  </head>

  <body style="font-family: Poppins, Verdana, Geneva, Tahoma, sans-serif;">
    <div id="main" style="width:100%;">
      <div data-server-rendered="true">
        <div style="background-color: #0b82c5; text-align:left;  padding:2%">
          <img width="120px" src="https://cdn.3xr.com/images/3xr-email.png" style="float: right; padding-right:20px;">
          <h1 style="color:#FFF;padding-left:8%; padding-top:20px; padding-bottom: 20px;">Revision Requested</h1>
        </div>
        <div style=" padding-right:10%; padding-left:10%; padding-top:20px;">
          <p>productName has been sent back to our artists with your feedback, we will
            notify you again when your requested revisions have been made. Please expect
            a 48 hour turnaround.</p>
          <p style="padding-top:15px; padding-bottom:15px;">
            <a href="http://www.product.link" style="color:#FFF; cursor:pointer;">
              <button style="cursor:pointer;background-color:#dda349; color=:#FFF; border:none; border-radius:5px;  padding-left:15px;  padding-right:15px; padding-top:5px; padding-bottom:5px; font-size:20px">Review asset</button>
            </a>
          </p>
          <div style="border-bottom:lightgray solid 2px"></div>
          <p>Have questions? Email
            <a href="mailto:sales@3xr.com" style="color: #0b82c5;">sales@3xr.com</a>
          </p>
        </div>
      </div>
    </div>
  </body>

</html>
`;
